CedarJava/src/main/java/com/cedarpolicy/model/slice/Policy.java

```diff
@@ -51,4 +65,33 @@ public Policy(
     public String toString() {
         return "// Policy ID: " + policyID + "\n" + policySrc;
     }
+
+    public static Policy parseStaticPolicy(String policyStr) throws InternalException, NullPointerException {
+        var policyText = parsePolicyJni(policyStr);
+        return new Policy(policyText, null);
+    }
+
+    public static Policy parsePolicyTemplate(String templateStr)  throws InternalException, NullPointerException {
+        var templateText = parsePolicyTemplateJni(templateStr);
+        return new Policy(templateText, null);
+    }
+
+    /**
+     * This method takes in a Policy and a list of Instantiations and calls Cedar JNI to ensure those slots
+     * can be used to instantiate the template. If the Template is validated ahead of time by using Policy.parsePolicyTemplate
+     * and the Instantiations are also ensured to be valid (for example, by validating their parts using EntityTypeName.parse
+     * and EntityIdentifier.parse), then this should only fail because the slots in the template don't match the instantiations
+     * (barring JNI failures).
+     * @param p Policy object constructed from a valid template. Best if built from Policy.parsePolicyTemplate
+     * @param principal EntityUid to put into the principal slot. Leave null if there's no principal slot
```

Comment:
Optional is not really a fit though, we need to surface why it failed if it fails. We have two choices that offer library ergonomics that are reasonable to java programmers: 
* Just throw exceptions
* Use a library that implements an either monad, or implement it ourselves

Reply:
Sorry, I should have been more clear. I'd prefer that the `principal` and `resource` be `Optional<EntityUID>` instead of using `null` when there's no slot. I agree we should throw an exception on failure, not return an `Optional<boolean>`.